### PR TITLE
OLMIS-7793: Changed AMC calculation for SM facilities

### PR DESCRIPTION
**Pull request by anawrotsoldevelo**
Jira link: https://openlmis.atlassian.net/browse/OLMIS-7793

- - -
This change is [Reviewable](https://reviewable.io/reviews/OpenLMIS/openlmis-requisition/93)

### CODE REVIEW

**Review by pwargulak — CHANGES_REQUESTED** [review]
I'll continue looking into the PR

File: src/main/java/org/openlmis/requisition/web/RequisitionV2Controller.java [hunk 1]
```diff
@@ -131,9 +137,32 @@ public RequisitionV2Dto updateRequisition(@PathVariable("id") UUID requisitionId
     logger.debug("Updating requisition with id: {}", requisitionId);
 
     profiler.start("UPDATE");
-    requisitionToUpdate.updateFrom(result.getRequisition(),
-        result.getOrderables(), result.getApprovedProducts(),
-        datePhysicalStockCountCompletedEnabledPredicate.exec(result.getProgram()));
+
+    if (requisitionToUpdate.getTemplate().isPopulateStockOnHandFromStockCards()) {
```
[the remark this answers is not in the document]
Comment (nawrotadam) [reply src/main/java/org/openlmis/requisition/web/RequisitionV2Controller.java:141]: I like your proposed approach with passing requisitionService to updateFrom. I will attempt to cleanup the code accordingly.

File: src/main/java/org/openlmis/requisition/domain/requisition/Requisition.java [hunk 2]
```diff
@@ -1074,6 +1164,31 @@ private void updateConsumptions(Map<VersionIdentityDto, OrderableDto> orderables
     }
   }
 
+  private void updateStockConsumptions(
+      Map<VersionIdentityDto, OrderableDto> orderables,
+      List<StockCardRangeSummaryDto> stockCardRangeSummariesToAverage,
+      List<ProcessingPeriodDto> periods) {
+
+    if (template.isColumnInTemplateAndDisplayed(ADJUSTED_CONSUMPTION)) {
```
Comment (pwargulak) [inline src/main/java/org/openlmis/requisition/domain/requisition/Requisition.java:1172]: This is more like FYI, question, not necessary a request to change.

This looks strange, `calculateAndSetStockBasedAverageConsumption` checks for `AVERAGE_CONSUMPTION` column too. If both columns are  displayed, it runs `getNonSkippedFullSupplyRequisitionLineItems` twice. 
Inside the 1st `if`, the value is calculated by some helper method and set inside RequisitionLineItem, but the 2nd `if` uses `RequisitionLineItem.calculateAndSetStockBasedAverageConsumption` method to calculate (so RequisitionLineItem sets its fields by itself) - this gives impression that `RequisitionLineItem` is a full OOP Object with methods and simple value-object at the same time, which is bad design IMO.

I was thinking about this suggestion, but it dosn't solve all concerns:
```
if(template.isColumnInTemplateAndDisplayed(ADJUSTED_CONSUMPTION) || template.isColumnInTemplateAndDisplayed(AVERAGE_CONSUMPTION))  {
  for(RequisitionLineItem line : getNonSkippedFullSupplyRequisitionLineItems(orderables)) {
    if(template.isColumnInTemplateAndDisplayed(ADJUSTED_CONSUMPTION)) {
      line.setAdjustedConsumption(...);
    }

    if(template.isColumnInTemplateAndDisplayed(AVERAGE_CONSUMPTION)) {
      StockCardRangeSummaryDto summaryToAverage = findStockCardRangeSummary(
            stockCardRangeSummariesToAverage, line.getOrderable().getId());

        line.calculateAndSetStockBasedAverageConsumption(summaryToAverage,
            template, periods, previousRequisitions);
    }
  }
}
```

**Review by pwargulak — APPROVED** [review]
Left some questions

**Pull request merged**